MediaRecorder shouldn't perform memory allocation on the thread it is receiving audio samples

jyavenard opened the pull request:
#### cb76829ab268a109558cca5efdce6ca18913e643
<pre>
MediaRecorder shouldn&apos;t perform memory allocation on the thread it is receiving audio samples
<a href="https://bugs.webkit.org/show_bug.cgi?id=282374">https://bugs.webkit.org/show_bug.cgi?id=282374</a>
<a href="https://rdar.apple.com/problem/138978311">rdar://problem/138978311</a>

Reviewed by Youenn Fablet.

We used to create a CMSampleBuffer in the audio thread before sending it to the AudioSampleBufferCompressor
for encoding.

We now instead allocate an InProcessCARingBuffer when we receive the first audio sample.
From that point, we push the new audio in the ringbuffer and send a message to the encoder&apos;s WorkQueue
to retrieve the data where the CMSampleBuffer will be created.
This reduces the number of memory allocation to once only.
The CARingBuffer is made to allow for 2s of audio. This is the same size used by the GPU process
to capture the audio and send it to the content process in its own ringbuffer.

No change in observable behaviour. Covered by existing tests.

* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::setSampleCount):
(WebCore::WebAudioBufferList::createWebAudioBufferListWithBlockBuffer):
(WebCore::WebAudioBufferList::setSampleCountWithBlockBuffer):
(WebCore::WebAudioBufferList::initializeList):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.h: We add a new method to create an AudioBufferList backed by a CMBlockBoffer.
This allows to directly copy the data from the CARingBuffer to the CMSampleBuffer without an intermediary.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesAvailableStarted):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesAvailable):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame): Fly-By: check if recorder is stopped before enqueuing. So that even in-flight video frames
at the time the recorder was stopped will be processed and converted.
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::maybeStartWriter):
(WebCore::MediaRecorderPrivateEncoder::stopRecording):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
(WebCore::MediaRecorderPrivateEncoder::hasAudio const): Deleted.
(WebCore::MediaRecorderPrivateEncoder::hasVideo const): Deleted.
(WebCore::MediaRecorderPrivateEncoder::WTF_GUARDED_BY_CAPABILITY): Deleted.

Canonical link: <a href="https://commits.webkit.org/286083@main">https://commits.webkit.org/286083@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ccee3969e209d074774273e7094510346337c8e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74801 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54231 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27618 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79225 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26039 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76918 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63367 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2016 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58756 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17037 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77868 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/48908 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64279 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39150 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46179 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/21776 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24372 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67352 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22121 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80714 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2119 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1271 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67016 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2267 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64297 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66309 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16461 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10249 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8403 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2084 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4871 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2112 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3033 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2119 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
```diff
@@ -99,13 +99,51 @@ void WebAudioBufferList::setSampleCount(size_t sampleCount)
+    return m_blockBuffer;
+}
+
+void WebAudioBufferList::initializeList(uint8_t* data, size_t channelLength, size_t totalLength)
```
Can we use a span instead of data and totalLength?

jyavenard replied:
done

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -276,12 +277,70 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
     // explicitly allow the following allocation(s).
     DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
 
+    if (!m_currentAudioSamplingRate) {
```
In theory, it should be possible for microphone to change of sample rate.
We could check that m_currentAudioSamplingRate != description.sampleRate().
Though I guess, the compressor might not like the change and this would give the risk of changing the ring buffer.
Maybe, when we detect that the description changes, we stop the recording or we at least log an error.

jyavenard replied:
testing has indicated that the sampling rate can't change; if it did quite a few audio listener would be wrong.
If you start with a stream at 24kHz and you switch to one at 48, it would be resampled to 24 before getting there.

The assumption that the rate never changed has been there ever since the MediaRecorder was added, why worry about it now ?

eric-carlson replied:
> testing has indicated that the sampling rate can't change

In that case an ASSERT can't hurt

jyavenard replied:
> > testing has indicated that the sampling rate can't change
> 
> In that case an ASSERT can't hurt

it's there already, in `MediaRecorderPrivateEncoder::appendAudioSampleBuffer` `ASSERT(!m_currentAudioSamplingRate || m_currentAudioSamplingRate == description.sampleRate(), "sampling rate should never change over a session");`

jyavenard replied:
I added a detection for sampling rate change, which will abort the recorder and force writing the data already obtained.

I will take on a follow up fix that make use of a resampler if needed.

jyavenard replied:
for the record, taking this offline with @youennf I could produce a case where the sampleRate changed, I first had to enable the settings to have the capture in the GPU process and then start recording and in the middle of the session do: 
```
setTimeout(() => {
            stream.getAudioTracks()[0].applyConstraints({sampleRate:44100})
        }, 1500);
```
it changed from 48kHz to 44.1kHz

jyavenard replied:
will follow up in https://bugs.webkit.org/show_bug.cgi?id=282489

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -276,12 +277,70 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
+        });
+    }
     ASSERT(!m_currentAudioSamplingRate || m_currentAudioSamplingRate == description.sampleRate(), "sampling rate should never change over a session");
     m_currentAudioSamplingRate = description.sampleRate();
```
Why not setting m_currentAudioSamplingRate in the above if check?

eric-carlson commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -276,12 +277,70 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
+    ASSERT(m_audioFormatDescription);
+
+    auto* absd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(m_audioFormatDescription.get());
+    ASSERT(absd);
```
Why not log and return early if the absd is null, as you do for all of the other error conditions?

jyavenard replied:
because it can't be here. we created the m_audioFormatDescription ourselves from an AudioBaseSampleDescription.
The only time absd will be null if the `AudioFormatDescription` wasn't from an audio object but a video object.

Hence the ASSERT.

eric-carlson replied:
If it "can't happen", why bother asserting? If it can happen, the we will crash in the dereference in a release build.

youennf commented on Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
```diff
@@ -99,13 +99,50 @@ void WebAudioBufferList::setSampleCount(size_t sampleCount)
+
+std::optional<std::pair<UniqueRef<WebAudioBufferList>, RetainPtr<CMBlockBufferRef>>> WebAudioBufferList::createWebAudioBufferListWithBlockBuffer(const CAAudioStreamDescription& description, size_t sampleCount)
+{
+    if (!sampleCount)
```
ASSERT(sampleCount) maybe?

youennf commented on Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
```diff
@@ -99,13 +99,50 @@ void WebAudioBufferList::setSampleCount(size_t sampleCount)
+
+    CMBlockBufferRef blockBuffer = nullptr;
+    if (auto error = PAL::CMBlockBufferCreateWithMemoryBlock(kCFAllocatorDefault, nullptr, bufferSizes->second, kCFAllocatorDefault, nullptr, 0, bufferSizes->second, kCMBlockBufferAssureMemoryNowFlag, &blockBuffer))
+        return { };
```
We usually log this error case elsewhere in our codebase.

jyavenard replied:
done

youennf commented on Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
```diff
@@ -99,13 +99,50 @@ void WebAudioBufferList::setSampleCount(size_t sampleCount)
+    RetainPtr block = adoptCF(blockBuffer);
+    char* data = { };
+    if (auto error = PAL::CMBlockBufferGetDataPointer(blockBuffer, 0, nullptr, nullptr, &data))
+        return { };
```
Ditto

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -276,21 +277,93 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
+        });
+    }
+    ASSERT(m_currentAudioSamplingRate == description.sampleRate(), "sampling rate should never change over a session");
+    if (m_currentAudioSamplingRate != description.sampleRate()) {
```
This seems ok as is. I wonder whether we would not want to check the whole description (so move from m_currentAudioSamplingRate to m_currentAduioDescription) maybe

jyavenard replied:
I do so in the follow up

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -276,21 +277,93 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
+            if (RefPtr protectedThis = weakThis.get())
+                protectedThis->stopRecording();
+        });
+        m_isStopped = true;
```
It is strange to do this assignment there.
That makes stopRecording() a no-op.
Maybe the above if check should be: `if (m_currentAudioSamplingRate != description.sampleRate() && !m_isStopped())` and remove this assignment here.

jyavenard replied:
All this will be gone in the follow up where we now handle change of format

jyavenard replied:
I'm going to continue to call `stopRecording()` and return early.

youennf reviewed (approved):
LGTM, but please check the `m_isStopped` assignment question

jyavenard committed:
MediaRecorder shouldn't perform memory allocation on the thread it is receiving audio samples
https://bugs.webkit.org/show_bug.cgi?id=282374
rdar://problem/138978311

Reviewed by Youenn Fablet.

We used to create a CMSampleBuffer in the audio thread before sending it to the AudioSampleBufferCompressor
for encoding.

We now instead allocate an InProcessCARingBuffer when we receive the first audio sample.
From that point, we push the new audio in the ringbuffer and send a message to the encoder's WorkQueue
to retrieve the data where the CMSampleBuffer will be created.
This reduces the number of memory allocation to once only.
The CARingBuffer is made to allow for 2s of audio. This is the same size used by the GPU process
to capture the audio and send it to the content process in its own ringbuffer.

No change in observable behaviour. Covered by existing tests.

* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::setSampleCount):
(WebCore::WebAudioBufferList::createWebAudioBufferListWithBlockBuffer):
(WebCore::WebAudioBufferList::setSampleCountWithBlockBuffer):
(WebCore::WebAudioBufferList::initializeList):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.h: We add a new method to create an AudioBufferList backed by a CMBlockBoffer.
This allows to directly copy the data from the CARingBuffer to the CMSampleBuffer without an intermediary.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesAvailableStarted):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesAvailable):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame): Fly-By: check if recorder is stopped before enqueuing. So that even in-flight video frames
at the time the recorder was stopped will be processed and converted.
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::maybeStartWriter):
(WebCore::MediaRecorderPrivateEncoder::stopRecording):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
(WebCore::MediaRecorderPrivateEncoder::hasAudio const): Deleted.
(WebCore::MediaRecorderPrivateEncoder::hasVideo const): Deleted.
(WebCore::MediaRecorderPrivateEncoder::WTF_GUARDED_BY_CAPABILITY): Deleted.

Canonical link: https://commits.webkit.org/286083@main
